src/net_processing.cpp

```diff
@@ -4423,6 +4443,24 @@ class CompareInvMempoolOrder
 };
 }
 
+bool PeerManagerImpl::SetupAddressRelay(CNode& node) const
+{
+    // We don't participate in addr relay with outbound block-relay-only
+    // connections to prevent providing adversaries with the additional
+    // information of addr traffic to infer the link.
+    if (node.IsBlockOnlyConn()) return false;
+
+    PeerRef peer = GetPeerRef(node.GetId());
+    if (!peer->addr_relay_enabled) {
+        // First addr message we have received from the peer, initialize
+        // m_addr_known
+        peer->m_addr_known = std::make_unique<CRollingBloomFilter>(5000, 0.001);
+        peer->addr_relay_enabled = true;
```

Comment:
You can atomically test and set the flag as follows:

```suggestion
    if (!peer->addr_relay_enabled.exchange(true)) {
        // First addr message we have received from the peer, initialize
        // m_addr_known
        peer->m_addr_known = std::make_unique<CRollingBloomFilter>(5000, 0.001);
```

(`std::atomic<T>::exchange()` returns the value of the object before the call (https://en.cppreference.com/w/cpp/atomic/atomic/exchange))

Reply:
nice trick! done